Order Syncing

braidn opened the pull request:
[Clubhouse Ticket](https://app.clubhouse.io/dynamomtl/story/48/)

More information:
https://github.com/glossier/solidus_retail/issues/7

braidn committed:
Import and Cleanup Order Sync Task.

This is a previous implementation in another app. Imported here and
cleaned up through linting with Rubocop. Both rake task and class. Tests
will come to encapsulate the logic in the Generate POS Order class.

GitHub:
https://github.com/glossier/solidus_retail/issues/7#issuecomment-246685062
Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Green Test For POS Orders

Correct a few naming issues and get the test suite green surrounding the
pos to order class.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Add Order Json Data For Mocking.

JSON data pulled from the mocks of the shopify_api gem. We will pull
this JSON in and mock the request to a specific Shopify endpoint and
this JSON will be returned and passed to the Shopify::Order.find return
when resolving.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Add In More Requirements For Testing Order Import

Require the shopify_api in the lib test along with adding in the shared
context that allows mocking of requests to the Shopify API.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn commented on spec/support/shared/shopify_shop.rb:
```diff
@@ -0,0 +1,20 @@
+    gateway
+  end
+
+  def mock_request(endpoint, extension)
```
Hmm @cabouffard and @DanielWright this resolves just fine. It creates the correct webmock, yadda yadda. However, I then want to invoke this mock by calling `ShopifyAPI::Order.find(order_numb)` but this fails with the following:

```
NoMethodError: undefined method `path' for nil:NilClass
from /Users/braidn/src/glossier-retail/.bundle/ruby/2.3.0/gems/activeresource-4.1.0/lib/active_resource/base.rb:688:in `prefix'
```

This occurs regardless of calling the mock_request method so it seems like it's a deeper issue. Any ideas would be amaze. Thanks.

DanielWright replied:
If you get a more complete backtrace, does that reveal anything interesting? I've seen errors like that when I was trying to retrieve a Transaction, because the correct syntax is `Transaction.find(id, params: { order_id: order_id })`. Otherwise, AResource didn't know how to build the full URL.

braidn replied:
Yeah, feel like I have found the error. It's totally down to AResource not knowing how to build the full URL. Maybe go it.

thisiscab replied:
It's weird though because when looking at the spec, the way you're calling the API seems to reflect how they do it in their testing suite. I don't think there is a param missing:

https://github.com/Shopify/shopify_api/blob/master/test/order_test.rb#L13

braidn commented on spec/support/shared/shopify_shop.rb:
```diff
@@ -0,0 +1,20 @@
+    gateway
+  end
+
+  def mock_request(endpoint, extension)
```
I feel like it may have something to do with the overall [test harness setup surrounding](http://stackoverflow.com/questions/10545162/undefined-method-path-for-nilnilclass-using-chargify-api-ares-gem) the Shopify configuration.

braidn commented:
OK, so maybe fiddled my way into some fixes to allow the ShopifyAPI mock request above to function properly.

braidn committed:
Successfully Stub Requests To Shopify API

Fix all stubbing woes. Remove all the logic to stub a request, including
the require to the shopify_api to a new context. Also add some simple
helpers that run in a before :all block to setup the ShopifyAPI properly
for the tests in each group.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn commented:
Nailed it (minus Travis) all. Thank you all for the help.

braidn committed:
Rake Task Updated With Correct Order Class

Rake task to pull orders from the Retail api to the Solidus store
updated with the correct class.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Passing POS Order Spec.

Arrive at a single passing test that calls out to a mock request for the
order return from Shopify. Green tests for this class from here on out.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Add POS Bool Migration For Spree Order

Migration for pos boolean on Spree Order added to the app.

braidn committed:
Add Pos Order Number To Spree Order

The order object get's updated with this attribute to link it to the
Shopify order name. Added migration

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
POS Order Creates A Correct Spree Order.

The order that comes from the point of sale can now successfully create
an order in the Solidus store. Some slight refactoring going on
throughout the pos order generator such as:

* Currency imported from the pos order
* Certain parts of the test can be skipped using the
deployed_environment? boolean

Green tests for the pos order spec.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn commented:
Seemingly having a little issue where the state machine is getting stuck at address and this is the error: `We are unable to calculate shipping rates for the selected items`. Currently hunting down the fix but, thought I would throw it in here to see if anyone has experienced this before?

cc: @cabouffard @DanielWright @bryanmtl

braidn commented:
Ok, maybe asked for help too quickly. We are having issues here: https://github.com/solidusio/solidus/blob/423fcb7920b356379af5ab98c44ed58481d153b4/core/app/models/spree/order.rb#L753

braidn commented:
Ok, so if I send that private (ensure_available_shipping_rates) method to our order it executes the code within the `if`. However if I check the logic of the `if` statement, both return false meaning it _should_ not execute the block... bizzaro.

braidn committed:
Use Shipping Method Method

Utilize the shipping method function to add a default shipping method to
the order shipments.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn commented:
I believe I might have found the issue here and as always it stems from my understanding of how these payment gateways are set up. @DanielWright will the new gateway have a source? Meaning we won't need to pass a source when creating a payment? This is how the current implementation is coded but, just questioning.

thisiscab commented:
From what I understand there will not be any source for the gateway we made!

braidn commented:
This is quite odd to me because, how can the app capture on this Gateway? There must be something special about that Gateway? @DanielWright do we have a factory for that gateway?

DanielWright commented:
We don't have a factory, no.

braidn committed:
Bump Shopify API Sig in Request Helper

One can not simply mock the Shopify API using webmock. Bump the api to
3.2 for the webmock user agent as to satisfy the mocking gods.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
Create Retail Gateway Factory

Simple gateway factory to have access to the Shopify Gateway class.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn commented:
We now do.

braidn commented:
@bryanmtl or @aamyot do you have an example setup with the current Shopify payment method? I am just trying to wrap my head around not requiring a source for the payments

braidn commented:
@DanielWright should [this be set to return a not implemented error?](https://github.com/glossier/solidus_retail/blob/master/app/models/spree/gateway/shopify_gateway.rb#L47)  I ask because the class [expects to call the `capture` method](https://github.com/glossier/solidus_retail/blob/feature/shopify-order-importer/lib/solidus_retail/order/generate_pos_order.rb#L112) on the payment.

thisiscab commented:
I'm the one who added this `not implemented error`. I thought that it was not necessary to implement that method and that if we ever had to use it then it would raise an error so we would know it's not implemented.

When I talked with @bryanmtl we came to the conclusion that capture was not necessary. We might want to return a nil value or something instead of throwing and exception

braidn commented:
@cabouffard I would imagine we need to have captured a payment for us to rewind this in terms of voiding/refunding. What cha think?

thisiscab commented:
But the capturing is done on Shopify, not Solidus. So we should not have to bother with that.

A person makes a transaction at the store using Shopify. We only want to reimburse or void it from Solidus, we don't want to be able to create new payments from Solidus for Shopify.

braidn commented:
Would we need to modify some payment locally as to display if that payment was in a purchased/voided/refunded state? If we just interact with the API, how do we tell the backoffice UI what state, what order or more specifically: the order's payment is in?

DanielWright commented:
@cabouffard We don't need to implement a _real_ capture method, but we need the gateway to respond to the `capture` message in the way that `Spree::Payment` objects expect. It's principle of least surprise. The alternative is creating the `Spree::Payment` in the completed state, and that introduces all kinds of gnarly regressions because we can't assume the state-machine callbacks can be ignored.

braidn committed:
Return A True Billing Response For Retail Gateway

Since the host app will not be doing payment capture. Return a true
billing response object so that the order can move forward and be
completed with a payment in the completed state.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
Process Complete Orders From Shopify

Orders from shopify can now be processed successfully into the Solidus
ecosystem. These orders will come through as complete, with a payment,
and line items that match those of Shopify.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
Add Some Gem Dependencies

Shopify and Webmock exist elsewhere in the rando forks of the project.
Bring them in here to mock requests with the Shopify api and return
proper Shopify::Order objects to pass to the new POS order creator.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Add Mock JSON Data.

Use mock json data from the shopify API gem to simulate a fake response
from the shopify server. We will then take this response and pass it
into the Shopify::Order.find method which will return an order object
that can be decomposed.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48/

braidn committed:
Double Definition For VCR/Webmock

After collision with rebase, these two items are defined twice in the
Gemfile. Removed and now defined only once.

braidn committed:
ActiveRecord Messy Double Error

While using receive_message_chain, RSpec must inject several doubles
into the test data. This causes a very deep merge! method to throw an
obtuse, odd error. Removing the use of message chain and rely on the
Shopify PaymentMethod to satisfy the finder method fixes this.

GitHub:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
Use Order Updater Prior To Complete

Order is not quite in a place where it is completable at this stage
without passing it through the updater. The tests that wrap this class
now throw green confetti everywhere.

GitHub:
https://app.clubhouse.io/dynamomtl/story/48

braidn committed:
Remove Phase2 Set Conditional

This conditional will end up being more specific to bundled products.
Removed for now.

Clubhouse:
https://app.clubhouse.io/dynamomtl/story/48

braidn commented:
There is quite a bit of code going on here @cabouffard and @DanielWright however, the tests are now officially green with the ability to merge with master. Travis is complaining about some rubocop issue but besides that it's passing.

_Context:_
Orders created on Shopify, sync (not including bundles) between Shopify and the Host app. This is done through a rather large class (place for refactoring now that we have tests surrounding it) and a rake task.

braidn commented:
Pinging @cabouffard and @DanielWright and maybe even @bryanmtl for a Code review here again. Thank you.

_Context:_
Orders created on Shopify, sync (not including bundles) between Shopify and the Host app. This is done through a rather large class (place for refactoring now that we have tests surrounding it) and a rake task.

thisiscab commented:
LGTM ! After Travis passes we should be good to go :)

braidn committed:
Remove Useless Rescue In Shared Context

These style of rescues are relatively useless and they make Rubocop a
very unhappy customer. Cleaned up for code cleanliness and to make all
the Cops happy.

braidn commented:
Travis is all cleaned up. Will pull this in a little later and give you a second Bryan to look at it if you have time.
